src/java/htsjdk/samtools/sra/SRALazyRecord.java

```diff
@@ -0,0 +1,941 @@
+import ngs.ReadIterator;
+import ngs.Read;
+import ngs.Fragment;
+import ngs.ErrorMsg;
+
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * Created by andrii.nikitiuk on 8/25/15.
+ */
+public class SRALazyRecord extends SAMRecord {
+    private static final Log log = Log.getInstance(SRALazyRecord.class);
+
+    private boolean isAligned;
+    private ReadCollection run;
+    private Object sraIterator;
+    private String sraReadId;
+    private String sraAlignmentId;
+    private int unalignedReadFragmentIndex = -1;
+
+
+    private Set<LazyField> initializedFields = new HashSet<LazyField>();
+    private Set<LazyFlag> initializedFlags = new HashSet<LazyFlag>();
+    private Set<LazyAttribute> initializedAttributes = new HashSet<LazyAttribute>();
+
+    private enum LazyField {
+        ALIGNMENT_START,
+        MAPPING_QUALITY,
+        REFERENCE_NAME,
+        CIGAR_STRING,
+        BASES,
+        QUALS,
+        MATE_ALIGNMENT_START,
+        MATE_REFERENCE_NAME,
+        INFERRED_INSERT_SIZE
+    }
+
+    private enum LazyFlag {
+        READ_NEGATIVE_STRAND,
+        READ_PAIRED,
+        NOT_PRIMARY_ALIGNMENT,
+        MATE_NEGATIVE_STRAND,
+        MATE_UNMAPPED,
+        FIRST_OF_PAIR,
+        SECOND_OF_PAIR
+    }
+
+    private enum LazyAttribute {
+        RG
+    }
+
+    private static class FlagGetter {
+        String getterName;
+        boolean canCallOnNotPaired;
+
+        FlagGetter(String getterName, boolean canCallOnAligned) {
+            this.getterName = getterName;
+            this.canCallOnNotPaired = canCallOnAligned;
+        }
+    }
+
+    private static class AttributeTagInfo {
+        LazyAttribute lazyAttribute;
+        String getterImplName;
+
+        AttributeTagInfo(LazyAttribute lazyAttribute, String getterImplName) {
+            this.lazyAttribute = lazyAttribute;
+            this.getterImplName = getterImplName;
+        }
+    }
+
+    private static Map<LazyFlag, FlagGetter> lazyFlagsGetters;
+    static
+    {
+        lazyFlagsGetters = new HashMap<LazyFlag, FlagGetter>();
+        lazyFlagsGetters.put(LazyFlag.READ_NEGATIVE_STRAND, new FlagGetter("getReadNegativeStrandFlag", true));
+        lazyFlagsGetters.put(LazyFlag.READ_PAIRED, new FlagGetter("getReadPairedFlag", true));
+        lazyFlagsGetters.put(LazyFlag.NOT_PRIMARY_ALIGNMENT, new FlagGetter("getNotPrimaryAlignmentFlag", true));
+        lazyFlagsGetters.put(LazyFlag.MATE_NEGATIVE_STRAND, new FlagGetter("getMateNegativeStrandFlag", false));
+        lazyFlagsGetters.put(LazyFlag.MATE_UNMAPPED, new FlagGetter("getMateUnmappedFlag", false));
+        lazyFlagsGetters.put(LazyFlag.FIRST_OF_PAIR, new FlagGetter("getFirstOfPairFlag", false));
+        lazyFlagsGetters.put(LazyFlag.SECOND_OF_PAIR, new FlagGetter("getSecondOfPairFlag", false));
+    }
```

Comment:
Can you refactor this so that the getter for each flag is stored in a field inside the enum itself instead of a separate `Map`?